truffle/src/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/Instrumenter.java

```diff
@@ -460,4 +460,7 @@ public void onLoad(LoadSourceSectionEvent event) {
      */
     public abstract ExecutionEventNode lookupExecutionEventNode(Node node, EventBinding<?> binding);
 
+    // XXXXXXXXX
+    public abstract CompilationState sampleCompilationState();
```

Comment:
Instead of this I'd possibly like to have a `CompilationEventListener` in the instrumentation API, like `GraalTruffleRuntimeListener`, which would be fixed to allow me to do all the work I want to track compilations.

Reply:
Yes that sounds like a better idea than exposing a concrete sampling. 

Here is a sketch of what I have in mind for that API:

```java

class Instrumenter {

  EventBinding attachCompilationEventListener(CompilationEventListener)

}


static abstract class CompilationInfo {

  CompilationInfo(){}

  public long getId() {...}
  public String getRootSimpleName() {...}
  public String getRootQualifiedName() {...}
  public int getTier() {...}
  public boolean isClone();
  public int getTruffleNodeCount();

}

static final class CompilationSuccessInfo extends CompilationInfo{

  public long getCodeSize() {...}
  public long getGraalNodeCount() {...}

}

static final class CompilationAbortInfo extends CompilationInfo{

  public boolean isTemporary(); // other temp bailout
  public boolean isPermanent(); // permanent bailout
  public boolean isFatal(); // internal error
  public String getReason();

}


interface CompilationEventListener {

void onQueued(Compilationinfo info)
void onDequeued(CompilationAbortInfo info)
void onStart(Compilationinfo info)
void onSuccess(CompilationSuccessInfo info)
void onFailed(CompilationAbortInfo info)
void onInvalidate(CompilationAbortInfo info,)

}
```

Note I don't think we should track deoptimizations in general. They are not necessarily related to compilations. Maybe we should have a separate listener for those events?
Is this something you definitely need?


we can use the same guaranteed order as in the compilation listener to document this:
```
 ( ((queue . unqueue))*
    . queue . started
    . (success)
      | ( failed)
    . invalidate )*
```

I suggest that in the default runtime we allow attaching events, but there are just never any events triggered.

Reply:
Ok I'll try something minimal like this and ask for further comment when I've got the basics working.